### PR TITLE
Fix SASL/SCRAM + add option for SASL IAM & add option to disable unauthenticated clients

### LINKED ISSUE #1761: msk_cluster - Cannot create a cluster w/ authentication sasl_scram
### Summary

When I do 

```- name: Create MSK cluster
  community.aws.msk_cluster:
    name: "{{ item.name }}"
....................................................................
    authentication:
      sasl_scram: true
....................................................................
```

i get an error:

```
"msg": "Failed to create kafka cluster: Parameter validation failed:\nInvalid type for parameter ClientAuthentication.Sasl.Scram, value: True, type: <class 'bool'>, valid types: <class 'dict'>"
```

Full traceback:
```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_community.aws.msk_cluster_payload_roovgjj3/ansible_community.aws.msk_cluster_payload.zip/ansible_collections/community/aws/plugins/modules/msk_cluster.py", line 484, in create_or_update_cluster
  File "/tmp/ansible_community.aws.msk_cluster_payload_roovgjj3/ansible_community.aws.msk_cluster_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/modules.py", line 354, in deciding_wrapper
    return retrying_wrapper(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/ansible_community.aws.msk_cluster_payload_roovgjj3/ansible_community.aws.msk_cluster_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/cloud.py", line 119, in _retry_wrapper
    return _retry_func(
           ^^^^^^^^^^^^
  File "/tmp/ansible_community.aws.msk_cluster_payload_roovgjj3/ansible_community.aws.msk_cluster_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/cloud.py", line 69, in _retry_func
    return func()
           ^^^^^^
  File "/home/.../.local/lib/python3.11/site-packages/botocore/client.py", line 530, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.../.local/lib/python3.11/site-packages/botocore/client.py", line 919, in _make_api_call
    request_dict = self._convert_to_request_dict(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.../.local/lib/python3.11/site-packages/botocore/client.py", line 990, in _convert_to_request_dict
    request_dict = self._serializer.serialize_to_request(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.../.local/lib/python3.11/site-packages/botocore/validate.py", line 381, in serialize_to_request
    raise ParamValidationError(report=report.generate_report())
botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid type for parameter ClientAuthentication.Sasl.Scram, value: True, type: <class 'bool'>, valid types: <class 'dict'>
failed: [localhost] (item={'name': '...-msk-dev-2', 'configuration': '...-msk-dev-conf', 'version': '2.8.1', 'nodes': 3, 'ebs_volume_gb': 256, 'enhanced_monitoring': 'PER_TOPIC_PER_BROKER', 'instance_type': 'kafka.t3.small', 'open_monitoring': {'jmx_exporter': False, 'node_exporter': True}, 'subnets': ['subnet-...', 'subnet-...', 'subnet-...'], 'security_groups': ['sg-...'], 'tags': {'Payer': '...'}}) => {
    "ansible_loop_var": "item",
    "boto3_version": "1.26.72",
    "botocore_version": "1.29.72",
    "changed": false,
    "invocation": {
        "module_args": {
            "access_key": null,
            "authentication": {
                "sasl_scram": true,
                "tls_ca_arn": null
            },
```

I can create the cluster w/o `authentication` but it creates an `Unauthenticated` cluster.

### Issue Type

Bug Report

### Component Name

msk_cluster

### Ansible Version

ansible [core 2.14.3]
  config file = .../infrastructure/ansible.cfg
  configured module search path = ['/home/.../.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.11/site-packages/ansible
  ansible collection location = /home/.../.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.11.2 (main, Feb  8 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)] (/usr/bin/python3)
  jinja version = 3.0.3
  libyaml = True
```


### Collection Versions

```
$ ansible-galaxy collection list
[DEPRECATION WARNING]: DEFAULT_GATHER_SUBSET option, the module_defaults keyword is a more generic version and can apply to all calls to the M(ansible.builtin.gather_facts) or 
M(ansible.builtin.setup) actions, use module_defaults instead. This feature will be removed from ansible-core in version 2.18. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.

# /home/.../.ansible/collections/ansible_collections
Collection        Version
----------------- -------
amazon.aws        5.4.0  
ansible.posix     1.5.1  
community.aws     5.4.0  
community.general 6.5.0  
community.mongodb 1.5.1  
community.mysql   3.6.0  

# /usr/lib/python3.11/site-packages/ansible_collections
Collection                    Version
----------------------------- -------
amazon.aws                    5.2.0  
ansible.netcommon             4.1.0  
ansible.posix                 1.5.1  
ansible.utils                 2.9.0  
ansible.windows               1.13.0 
arista.eos                    6.0.0  
awx.awx                       21.12.0
azure.azcollection            1.14.0 
check_point.mgmt              4.0.0  
chocolatey.chocolatey         1.4.0  
cisco.aci                     2.4.0  
cisco.asa                     4.0.0  
cisco.dnac                    6.6.3  
cisco.intersight              1.0.23 
cisco.ios                     4.3.1  
cisco.iosxr                   4.1.0  
cisco.ise                     2.5.12 
cisco.meraki                  2.15.1 
cisco.mso                     2.2.1  
cisco.nso                     1.0.3  
cisco.nxos                    4.1.0  
cisco.ucs                     1.8.0  
cloud.common                  2.1.2  
cloudscale_ch.cloud           2.2.4  
community.aws                 5.2.0  
community.azure               2.0.0  
community.ciscosmb            1.0.5  
community.crypto              2.11.0 
community.digitalocean        1.23.0 
community.dns                 2.5.1  
community.docker              3.4.2  
community.fortios             1.0.0  
community.general             6.4.0  
community.google              1.0.0  
community.grafana             1.5.4  
community.hashi_vault         4.1.0  
community.hrobot              1.7.0  
community.libvirt             1.2.0  
community.mongodb             1.5.1  
community.mysql               3.6.0  
community.network             5.0.0  
community.okd                 2.3.0  
community.postgresql          2.3.2  
community.proxysql            1.5.1  
community.rabbitmq            1.2.3  
community.routeros            2.7.0  
community.sap                 1.0.0  
community.sap_libs            1.4.0  
community.skydive             1.0.0  
community.sops                1.6.1  
community.vmware              3.4.0  
community.windows             1.12.0 
community.zabbix              1.9.2  
containers.podman             1.10.1 
cyberark.conjur               1.2.0  
cyberark.pas                  1.0.17 
dellemc.enterprise_sonic      2.0.0  
dellemc.openmanage            6.3.0  
dellemc.os10                  1.1.1  
dellemc.os6                   1.0.7  
dellemc.os9                   1.0.4  
dellemc.powerflex             1.5.0  
dellemc.unity                 1.5.0  
f5networks.f5_modules         1.22.1 
fortinet.fortimanager         2.1.7  
fortinet.fortios              2.2.2  
frr.frr                       2.0.0  
gluster.gluster               1.0.2  
google.cloud                  1.1.2  
grafana.grafana               1.1.1  
hetzner.hcloud                1.10.0 
hpe.nimble                    1.1.4  
ibm.qradar                    2.1.0  
ibm.spectrum_virtualize       1.11.0 
infinidat.infinibox           1.3.12 
infoblox.nios_modules         1.4.1  
inspur.ispim                  1.3.0  
inspur.sm                     2.3.0  
junipernetworks.junos         4.1.0  
kubernetes.core               2.4.0  
lowlydba.sqlserver            1.3.1  
mellanox.onyx                 1.0.0  
netapp.aws                    21.7.0 
netapp.azure                  21.10.0
netapp.cloudmanager           21.22.0
netapp.elementsw              21.7.0 
netapp.ontap                  22.3.0 
netapp.storagegrid            21.11.1
netapp.um_info                21.8.0 
netapp_eseries.santricity     1.4.0  
netbox.netbox                 3.11.0 
ngine_io.cloudstack           2.3.0  
ngine_io.exoscale             1.0.0  
ngine_io.vultr                1.1.3  
openstack.cloud               1.10.0 
openvswitch.openvswitch       2.1.0  
ovirt.ovirt                   2.4.1  
purestorage.flasharray        1.17.0 
purestorage.flashblade        1.10.0 
purestorage.fusion            1.3.0  
sensu.sensu_go                1.13.2 
splunk.es                     2.1.0  
t_systems_mms.icinga_director 1.32.0 
theforeman.foreman            3.9.0  
vmware.vmware_rest            2.2.0  
vultr.cloud                   1.7.0  
vyos.vyos                     4.0.0  
wti.remote                    1.0.4  

```


### AWS SDK versions

```
$ pip show boto boto3 botocore
Name: boto
Version: 2.49.0
Summary: Amazon Web Services Library
Home-page: https://github.com/boto/boto/
Author: Mitch Garnaat
Author-email: mitch@garnaat.com
License: MIT
Location: /usr/local/lib/python3.11/site-packages
Requires: 
Required-by: 
---
Name: boto3
Version: 1.26.72
Summary: The AWS SDK for Python
Home-page: https://github.com/boto/boto3
Author: Amazon Web Services
Author-email: 
License: Apache License 2.0
Location: /home/.../.local/lib/python3.11/site-packages
Requires: botocore, jmespath, s3transfer
Required-by: 
---
Name: botocore
Version: 1.29.72
Summary: Low-level, data-driven core of boto 3.
Home-page: https://github.com/boto/botocore
Author: Amazon Web Services
Author-email: 
License: Apache License 2.0
Location: /home/.../.local/lib/python3.11/site-packages
Requires: jmespath, python-dateutil, urllib3
Required-by: awscli, boto3, s3transfer

```


### Configuration

```$ ansible-config dump --only-changed
[DEPRECATION WARNING]: DEFAULT_GATHER_SUBSET option, the module_defaults keyword is a more generic version and can apply to all calls to the M(ansible.builtin.gather_facts) or 
M(ansible.builtin.setup) actions, use module_defaults instead. This feature will be removed from ansible-core in version 2.18. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
CACHE_PLUGIN(/home/.../infrastructure/ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/home/.../infrastructure/ansible.cfg) = ~/.ansible/cache
CACHE_PLUGIN_TIMEOUT(/home/.../infrastructure/ansible.cfg) = 3600
CALLBACKS_ENABLED(/home/.../infrastructure/ansible.cfg) = ['timer', 'profile_tasks', 'profile_roles']
CONFIG_FILE() = /home/.../infrastructure/ansible.cfg
DEFAULT_ASK_PASS(/home/.../infrastructure/ansible.cfg) = False
DEFAULT_EXECUTABLE(/home/.../infrastructure/ansible.cfg) = /bin/bash
DEFAULT_FORCE_HANDLERS(/home/.../infrastructure/ansible.cfg) = True
DEFAULT_FORKS(/home/.../infrastructure/ansible.cfg) = 15
DEFAULT_GATHERING(/home/.../infrastructure/ansible.cfg) = smart
DEFAULT_GATHER_SUBSET(/home/.../infrastructure/ansible.cfg) = ['all']
DEFAULT_HOST_LIST(/home/.../infrastructure/ansible.cfg) = ['/home/.../infrastructure/envs']
DEFAULT_LOG_PATH(/home/.../infrastructure/ansible.cfg) = /home/.../.ansible/ansible.log
DEFAULT_MANAGED_STR(/home/.../infrastructure/ansible.cfg) = Ansible managed! DON'T CHANGE THIS FILE BY HAND! You were warned!
DEFAULT_ROLES_PATH(/home/.../infrastructure/ansible.cfg) = ['/home/.../infrastructure/roles']
DEFAULT_TIMEOUT(/home/.../infrastructure/ansible.cfg) = 30
DEPRECATION_WARNINGS(/home/.../infrastructure/ansible.cfg) = True
HOST_KEY_CHECKING(/home/.../infrastructure/ansible.cfg) = False
INVENTORY_ENABLED(/home/.../infrastructure/ansible.cfg) = ['yaml', 'aws_ec2', 'ini']
RETRY_FILES_ENABLED(/home/.../infrastructure/ansible.cfg) = False
SHOW_CUSTOM_STATS(/home/.../infrastructure/ansible.cfg) = True


```

### OS / Environment

Fedora release 37 (Thirty Seven)

### Steps to Reproduce

```
- name: Create MSK cluster
  community.aws.msk_cluster:
    name: "{{ item.name }}"
....................................................................
    authentication:
      sasl_scram: true
```

### Expected Results

Create a MSK cluster w/ SASL/SCRAM authentication.

### Actual Results

```
"msg": "Failed to create kafka cluster: Parameter validation failed:\nInvalid type for parameter ClientAuthentication.Sasl.Scram, value: True, type: <class 'bool'>, valid types: <class 'dict'>"
```

As I'm working on this I can jump in at any time to further debug this.

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by eRadical**
##### SUMMARY
- fix SASL/SCRAM - Fixes #1761 
- add IAM authentication
- add option to disable unauthenticated clients

Many thanx to @markuman for throwing me into this issue.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
msk_cluster

##### ADDITIONAL INFORMATION
I will probably add more tests after working w/ this.

### CODE REVIEW

**Comment by tremble** [comment]
@eRadical I see you fighting with the formatting/darker stuff

If you install "darker" ( https://pypi.org/project/darker/ ) it should be possible to just run the darker command in the `community/aws/` directory and it'll apply all the formatting for you (including some of the tweaks we have, like line length)

File: plugins/modules/msk_cluster.py [hunk 1]
```diff
@@ -123,6 +123,14 @@
                 description: SASL/SCRAM authentication is enabled or not.
                 type: bool
                 default: False
+            sasl_iam:
+                description: IAM authentication is enabled or not.
+                type: bool
+                default: False
```
Comment (tremble) [inline plugins/modules/msk_cluster.py:129]:
```suggestion
                default: False
                version_added: 5.5.0
```

**Review by tremble — CHANGES_REQUESTED** [review]
Many thanks for this.

Down at line 695 there's a "module_args" parameter, this will also need to be updated, the DOCUMENTATION fragment is only documentation (upstream Ansible is aware that this isn't the greatest, but changing this for modules is non-trivial)

Does it make sense for all of these new options to have default values?  Will they conflict with each other?  it might make more sense for 
```
msk_cluster:
   ...
   authentication: {}
   ...
```

To be the way to disable authentication, rather than an explicitly parameter (especially one with a default value.

File: changelogs/fragments/1761-msk_cluster-fix-authentication.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+bugfixes:
+- msk_cluster - fix creating a cluster w/ sasl_scram authentication (https://github.com/ansible-collections/community.aws/issues/1761).
```
[the remark this answers is not in the document]
Comment (eRadical) [reply changelogs/fragments/1761-msk_cluster-fix-authentication.yml:2]: will do

**Comment by eRadical** [comment]
> To be the way to disable authentication, rather than an explicitly parameter (especially one with a default value.

You need to explicitly disable "Unauthenticated access".

It looks like this:
image

As you can see there are 4 independed checkboxes.

**Comment by markuman** [comment]
> When I created a cluster w/ Ansible it created only w/ "Unauthenticated" checked. So I suspect the boto default is actually "Unauthenticated".

For us it is the default then. Otherwise it would break backwards compatibility.

File: plugins/modules/msk_cluster.py [hunk 3]
```diff
@@ -123,6 +123,15 @@
...
+                version_added: 5.5.0
+            unauthenticated:
+                description: Option to explicitly turn on or off authentication
```
Comment (markuman) [inline plugins/modules/msk_cluster.py:132]:
```suggestion
                description: Option to explicitly turn on or off authentication
```

File: plugins/modules/msk_cluster.py [hunk 4]
```diff
@@ -123,6 +123,15 @@
...
+                type: bool
+                default: True
```
Comment (markuman) [inline plugins/modules/msk_cluster.py:134]:
```suggestion
            unauthenticated:
                version_added: 5.5.0
                description: Option to explicitly turn on or off authentication
                type: bool
                default: True
```

**Comment by markuman** [comment]
> ERROR: plugins/modules/msk_cluster.py:132:80: W291: trailing whitespace

 is fixed and committed.

> ERROR: plugins/modules/msk_cluster.py:0:0: doc-default-does-not-match-spec: Argument 'sasl_iam' in argument_spec found in authentication defines default as (None) but documentation defines default as (False)
> ERROR: plugins/modules/msk_cluster.py:0:0: doc-default-does-not-match-spec: Argument 'unauthenticated' in argument_spec found in authentication defines default as (None) but documentation defines default as (True)

```diff
diff --git a/plugins/modules/msk_cluster.py b/plugins/modules/msk_cluster.py
index 31e2fe6a..46681d95 100644
--- a/plugins/modules/msk_cluster.py
+++ b/plugins/modules/msk_cluster.py
@@ -126,12 +126,10 @@ options:
             sasl_iam:
                 description: IAM authentication is enabled or not.
                 type: bool
-                default: False
                 version_added: 5.5.0
             unauthenticated:
                 description: Option to explicitly turn on or off authentication
                 type: bool
-                default: True
     enhanced_monitoring:
         description: Specifies the level of monitoring for the MSK cluster.
         choices:

```

sanity is surviving here locally. `ansible-test sanity  --python 3.10 plugins/modules/msk_cluster.py --docker default`  
but I haven't look deeper in the logic.

As mentioned above, we must keep backwards compatibility. So the module must result in the same settings, when those parameters are not set on execution.

**Comment by eRadical** [comment]
@markuman 

Can I do default & required?

```
    authentication:
        description:
            - Includes all client authentication related information.
            - Effective only for new cluster and can not be updated.
        type: dict
        suboptions:
            tls_ca_arn:
                description: List of ACM Certificate Authority ARNs.
                type: list
                elements: str
                default: False
                required: False
            sasl_scram:
                description: SASL/SCRAM authentication is enabled or not.
                type: bool
                default: False
                required: False
            sasl_iam:
                description: IAM authentication is enabled or not.
                type: bool
                default: False
                required: False
                version_added: 5.5.0
            unauthenticated:
                version_added: 5.5.0
                description: Option to explicitly turn on or off authentication
                type: bool
                default: True
                required: False
```
Note in `unauthenticated` that default is True and not required :)

**Comment by eRadical** [comment]
Nvm, I just did a push, let's see how it goes!

**Comment by markuman** [comment]
recheck

**Comment by markuman** [comment]
hm, maybe the entire msk test takes to long now. Any ideas @tremble?

**Comment by eRadical** [comment]
Yes, it might be. As they are now: tests are starting and stopping 4 clusters:
- old usual test,
- sasl_scram,
- sasl_iam,
- a cluster w/ sasl_scram  but `unauthenticated` disabled.

For me it takes about 1h to do a cluster change (switch unauthenticated to off or add sasl_iam).
I'll merge all the other tests into one tomorrow.

**Comment by eRadical** [comment]
Done!
I've also rebased.

**Comment by eRadical** [comment]
```
2023-04-01 08:10:33.996790 \| controller \| TASK [msk_cluster : update msk cluster] ****************************************
--
5100 | 2023-04-01 08:10:33.996795 \| controller \| task path: /home/zuul-worker/.ansible/collections/ansible_collections/community/aws/tests/integration/targets/msk_cluster/tasks/test_update.yml:24
5101 | 2023-04-01 08:12:10.438294 \| RUN END RESULT_TIMED_OUT: [untrusted : github.com/ansible/ansible-zuul-jobs/playbooks/ansible-test-base/run.yaml@master]
```

It times out at update so we're not reaching the second cluster, not even cluster deletion.

**Comment by markuman** [comment]
> It times out at update so we're not reaching the second cluster, not even cluster deletion.

cluster deletion is handled in that case the [aws-terminator](https://github.com/mattclay/aws-terminator).  

@tremble any ideas?

**Comment by markuman** [comment]
> hm, maybe the entire msk test takes to long now. Any ideas @tremble?

@goneri any ideas here?

**Comment by eRadical** [comment]
Same timeout.

(I pretty sure we can increase the timeout.)
Should we increase timeout?

**Comment by markuman** [comment]
recheck

**Comment by markuman** [comment]
> 2023-04-18 13:48:04.265593 | controller | botocore.errorfactory.ConflictException: An error occurred (ConflictException) when calling the CreateConfiguration operation: A resource with this name already exists.

Unclear for me if just the integration test has a bug or if this is a side effect by splitting the test.  
Any thoughts about it @tremble @eRadical?

**Comment by eRadical** [comment]
I'll test tomorrow in the office as I also need to start the prod clusters.

**Comment by tremble** [comment]
> > 2023-04-18 13:48:04.265593 | controller | botocore.errorfactory.ConflictException: An error occurred (ConflictException) when calling the CreateConfiguration operation: A resource with this name already exists.
> 
> Unclear for me if just the integration test has a bug or if this is a side effect by splitting the test. Any thoughts about it @tremble @eRadical?

The actual failure is:
```
2023-04-18 13:46:43.316236 | controller | TASK [msk_cluster-auth : validate return values] *******************************
2023-04-18 13:46:43.316241 | controller | task path: /home/zuul-worker/.ansible/collections/ansible_collections/community/aws/tests/integration/targets/msk_cluster-auth/tasks/test_create_auth.yml:49
2023-04-18 13:46:43.398248 | controller | fatal: [testhost]: FAILED! => {
2023-04-18 13:46:43.398293 | controller |     "msg": "The conditional check 'msk_cluster.cluster_info.client_authentication.sasl.scram.enabled == true' failed. The error was: error while evaluating conditional (msk_cluster.cluster_info.client_authentication.sasl.scram.enabled == true): 'dict object' has no attribute 'scram'. 'dict object' has no attribute 'scram'"
2023-04-18 13:46:43.398300 | controller | }
```

The cluster is then being left in the "deleting" state, and when it's retried (with added verbosity) it fails because the deletion hadn't completed.

**Comment by tremble** [comment]
@eRadical it looks like the failing test was caused by a combination of 2 things:
- AWS APIs not always returning disabled options (eg Unauthenticated)
- Enabling "Iam" auth was replacing enabling "SCRAM" auth

**Review by tremble — APPROVED** [review]
Note: for 5.5 we'll need to explicitly bump and test for the botocore versions

**Comment by tremble** [comment]
regate

**Comment by tremble** [comment]
@eRadical,

Thanks for your work on this.  Your work should be available when we release 5.5.0.  This is expected within the next couple of weeks (along with 6.0.0)

**Comment by eRadical** [comment]
@tremble - It was my pleasure! ...and (I think) also the normal path after working & promoting Ansible for a long time. I just hope it's just the beginning.

**Pull request merged**